torch/jit/__init__.py

```diff
@@ -1272,10 +1272,15 @@ def interface(obj):
         raise RuntimeError("interface must be applied to a class")
     if not _is_new_style_class(obj):
         raise RuntimeError("TorchScript interfaces must inherit from 'object'")
+
     qualified_name = _qualified_name(obj)
-    ast = get_jit_class_def(obj, obj.__name__)
     rcb = _jit_internal.createResolutionCallbackFromFrame(1)
-    torch._C._jit_script_interface_compile(qualified_name, ast, rcb)
+    # if this type is a `nn.Module` subclass, generate an module interface type
+    # instead of a class interface type, an module interface type only compile
+    # the user provided methods as part of the interface
+    is_module_interface = issubclass(obj, torch.nn.Module)
+    ast = get_jit_class_def(obj, obj.__name__)
+    torch._C._jit_script_interface_compile(qualified_name, ast, rcb, is_module_interface)
```

Comment:
we actually just pass a flag into `_jit_script_interface_compile` and it will just create an InterfaceType with is_module inside.

Reply:
I mean: instead of a flag, can we put the module information in the ClassDef itself? That avoids the need for an extra flag

Reply:
Agreed that ClassDef be the source of truth is better. I did some investigation on this approach, I am surprised that our python frontend now actually does not parse the `superclass` at all, it will leave the `Maybe<Expr>` as just a Maybe with no Expr inside. But string frontend does have it since it needs to support deserialization. Getting python frontend to parse base class should not be too hard but it's a fair amount of changes that is not related to this PR (we need to inspect the `mro()` and get all base classes someway). I will make a separate PR to support this.